### ISSUE TITLE
Iteration speed for V2W 2B post-training

### ISSUE THREAD

**Issue by VitorGuizilini-TRI** [nvidia-cosmos/cosmos-predict2#4]
Hi, thank you very much for the new repository! I am running post-training for the 2B V2W model on 8 x 80 GB GPUs, what's the expected iteration speed? I am getting roughly 80 seconds, doesn't seem correct to me.

**Comment by MaciejBalaNV** [comment]
Hey @VitorGuizilini-TRI, thanks for raising this issue. Could you specify which GPUs you're using? Are you using the docker image or conda environment?

**Comment by MaciejBalaNV** [comment]
@VitorGuizilini-TRI 
The NeMo dataset examples has only 4 videos in it. When we're using 8 GPUs, the dataset is consumed after each iteration and the iterators need to be refreshed, which causes a few seconds of overhead. That's why the 8 GPUs case is longer by 6s here. I'd guess that even with 1 GPU you are going to see a longer iteration time every couple of iterations as the dataset is exhaused. Even passing `persistent_workers=True` to the `DataLoader` doesn't help here. With bigger datasets this overhead is not significant. However, perhaps we should change the number of GPUs used in this example, as it indeed can be confusing.

A different thing that's worrying for me is the 80s iteration time in the baseline case that you reported. The speed I'm getting with the `nvcr.io/nvidia/cosmos/cosmos-predict2-container:1.0` docker image is around 30s, which is the expected value here. Let me look into the conda environment setup for A100 more closely.

**Comment by VitorGuizilini-TRI** [comment]
Thanks, I'll try the docker environment here as well. 
Just to make sure I understand: when using 8 GPUs, isn't the same model shared across all of them? So I'd be consuming samples at the same rate as if I'm using a single GPU, no? i.e. each sample is repeated across all GPUs each time, so they are exposed to the same observation.

**Comment by MaciejBalaNV** [comment]
Both CP and distributed sampling for data-parallelism is handled automatically in this codebase. CP is used mainly for saving memory. If you find that you don't need it for a specific resolution and GPU memory available, you can turn it off by modifying the training config. For the nemo example it's defined in `cosmos_predict2/configs/base/experiment/cosmos_nemo_assets.py` file as 
```
    model_parallel=dict(
        context_parallel_size=2,
    ),
```
setting it to 1 would disable CP.

**Comment by VitorGuizilini-TRI** [comment]
I'm getting this error when running the 2B V2W example (no modifications)
```
ERROR: Unexpected bus error encountered in worker. This might be caused by insufficient shared memory (shm).
```

**Comment by VitorGuizilini-TRI** [comment]
That's using the NeMo dataset, do you think if I switch to a larger dataset I'll get the 30s you mentioned? Or are you getting 30s with the same dataset?
Using a single GPU (no CP) and resolution 256x336 I'm getting ~2s (it oscillates between 1.5s and 3s, as you said)

**Comment by MaciejBalaNV** [comment]
@VitorGuizilini-TRI 
24s sounds good, I believe the 40s in your 8 GPUs training also comes from the dataset refresh overhead, so with a bigger dataset you should see stable ~24s iteration speed with 8 GPUs.

In order to enable gradient accumulation you need to modify the `trainer.grad_accum_iter` config value to an integer representing the number of iterations you want to accumulate gradient over. You can either do that in config specified in Python files, like in `cosmos_predict2/configs/base/experiment/cosmos_nemo_assets.py` or via command line by specifying `trainer.grad_accum_iter=<N>` after the ` -- ` separator.

**Comment by VitorGuizilini-TRI** [comment]
Thanks! Quick follow up, is it straightforward to increase batch size with a custom dataset? Is it just about changing the dataloader parameter, and the rest should work accordingly?

**Comment by MaciejBalaNV** [comment]
@VitorGuizilini-TRI We recently added a new post-training example in [this PR](https://github.com/nvidia-cosmos/cosmos-predict2/pull/26). There's also a new PR that adds the expected iteration speed for this example [here](https://github.com/nvidia-cosmos/cosmos-predict2/pull/36).

Closing this issue for now, as all questions are answered. Feel free to reopen in case of any new questions.
